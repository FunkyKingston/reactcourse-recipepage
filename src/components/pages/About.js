import React from 'react';
// import '../../styles/Pages.scss'; // enough to import via App.scss

const About = () => {
  return (
    <div className="content-area">

      <div className="section-line">
        <span className="section-header">ABOUT</span>
      </div>

      <div className="about">
        <h3>React course!</h3>
        <p>
          This first version of the page is for completing a task in a React course. 
          It uses the rich <a href="https://spoonacular.com/food-api/" target="_blank" rel="noopener noreferrer">Spoonacular Food API</a>. 
          Certain elements have been "forced into" the page to satisfy examination criteria.
          Being a huge food lover with a passion for sharing and having good times together, I actually have loose plans to make a 
          complete recipes page, with the specific focus to share recipes, menus and food-related events among friends. 
          That means designing my own little API, backend, authentication, including functionality for adding recipes, creating menus, etc.
        </p>
      </div>
      
    </div>
  )
}

export default About;