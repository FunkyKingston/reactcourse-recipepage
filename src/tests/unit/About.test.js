import React from 'react';
// import ReactDOM from 'react-dom'; // We don't use ReactDOM.render() here - the render functions provided by testing libraries are developed fit for purpose
import { BrowserRouter as Router } from 'react-router-dom';

// *** TESTING REACT APPS *** - https://jestjs.io/docs/en/tutorial-react
// Jest is a testing framework created and maintained by Facebook. If you build your React application with create-react-app, you can start 
// using Jest with zero config. 
// - Use Jest (provides a "test runner") with react-test-renderer and @testing-library/react library to conduct snapshot and DOM testing.

// *** React Test Renderer *** (Used e.g. for Snapshot tests) - https://reactjs.org/docs/test-renderer.html#overview 
// - not included with create-react-app -> $ npm i react-test-renderer -D
// - "React Test Renderer is a renderer for testing React Components, which traverses a ReactElement tree and represents as a JSON object. 
//    The renderer is combined with Jestâ€™s snapshot testing."
import TestRenderer from 'react-test-renderer'; // https://reactjs.org/docs/test-renderer.html

// *** React Testing Library *** (Used e.g. for DOM testing)
// - "introducing React Testing Library (Kent C. Dodds)" - https://kentcdodds.com/blog/introducing-the-react-testing-library
import { render, screen, fireEvent } from '@testing-library/React'; // https://testing-library.com/, https://testing-library.com/docs/react-testing-library/api
// * import userEvent from '@testing-library/user-event'; // Package specifically for testing user events - builds on top of the fireEvent API.
// * "React Testing Library re-exports everything from DOM Testing Library as well as these methods: ..." 
//   - https://testing-library.com/docs/dom-testing-library/api-queries/
//     - DOM Testing Library: https://testing-library.com/docs/dom-testing-library/api-queries

// Import About-component to test
import About from '../../components/pages/About';


///////////////////
// First example // - "Bad testing philosophy" - according to React Testing Library author
///////////////////
test("There exists a div element with className content-area", () => {
  render(<About />);
  // Getting a DOM element by className (like below) can be done with javascript, but NOT with react testing library, 
  // as it's not good testing philosophy to test implementation details - instead test the end result
  const divElement = document.getElementsByClassName("content-area"); 
  expect(divElement).not.toBeNull();
});


///////////////////////////////////////////////////////////////
// Example of React Testing Library - (super basic) DOM test // - https://www.robinwieruch.de/react-testing-library
///////////////////////////////////////////////////////////////
test("There exists an element that displays ABOUT", () => {
  render(<About />);
  // screen.debug() // debug whole document. debug() - "essentially a shortcut for console.log(prettyDOM())"
  // screen.debug(element) // debug single element

  const headerElement = screen.queryByText("ABOUT");
  expect(headerElement).not.toBeNull();
  const headerElement2 = screen.queryByText("React course!");
  expect(headerElement2).not.toBeNull();
});

// getBy vs. queryBy vs. findBy - https://testing-library.com/docs/dom-testing-library/api-queries
// - getBy* queries return the first matching node for a query, and throw an error if no elements match
// - queryBy* queries return the first matching node for a query, and return null if no elements match
// - findBy* queries return a promise which resolves when an element is found which matches the given query. 
//           The promise is rejected if no element is found or if more than one element is found after a default timeout of 1000ms

////////////////////////////////////
// Example of react-test-renderer //
////////////////////////////////////
// it() vs test() - The Jest docs state "it" is an alias of "test". So they are exactly the same. - https://jestjs.io/docs/en/api.html#testname-fn-timeout
it("Matches snapshot", () => {
  const tree = TestRenderer.create(<About />).toJSON(); // -> a virtual DOM _object_
  expect(tree).toMatchSnapshot(); // tree is compared to previously saved snapshot in the __snapshots__-folder (created the first time the test runs)
});


