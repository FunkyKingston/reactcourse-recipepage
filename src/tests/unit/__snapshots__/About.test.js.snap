// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<div
  className="content-area"
>
  <div
    className="section-line"
  >
    <span
      className="section-header"
    >
      ABOUT
    </span>
  </div>
  <div
    className="about"
  >
    <h3>
      React course!
    </h3>
    <p>
      This first version of the page is for completing a task in a React course. It uses the rich 
      <a
        href="https://spoonacular.com/food-api/"
        rel="noopener noreferrer"
        target="_blank"
      >
        Spoonacular Food API
      </a>
      . Certain elements have been "forced into" the page to satisfy examination criteria. Being a huge food lover with a passion for sharing and having good times together, I actually have loose plans to make a complete recipes page, with the specific focus to share recipes, menus and food-related events among friends. That means designing my own little API, backend, authentication, including functionality for adding recipes, creating menus, etc.
    </p>
  </div>
</div>
`;
